test/test_api.py

```diff
@@ -0,0 +1,220 @@
+
+
+@pytest.mark.parametrize("context", contexts)
+def test_as_contextmanager(context: str) -> None:
+    """
+    Expects
+    -------
+    * Should be able to use as a context manager
+    """
+    with Pynisher(subfunction, context=context) as restricted_func:
+        other_process_id = restricted_func()
+
+    this_process_id = os.getpid()
+    assert this_process_id != other_process_id
+
+
+@pytest.mark.parametrize("context", contexts)
+def test_call(context: str) -> None:
+    """
+    Expects
+    -------
+    * Should be able to call the restricted function
+    """
+    restricted_func = Pynisher(subfunction, context=context)
+
+    this_process_id = os.getpid()
+    other_process_id = restricted_func()
+
+    assert this_process_id != other_process_id
+
+
+@pytest.mark.parametrize("context", contexts)
+def test_run(context: str) -> None:
+    """
+    Expects
+    -------
+    * Should be able to explicitly call run on the restricted function
+    """
+    pynisher = Pynisher(subfunction, context=context)
+
+    this_process_id = os.getpid()
+    other_process_id = pynisher.run()
+    assert this_process_id != other_process_id
+
+
+def test_limit_usage() -> None:
+    """
+    Expects
+    -------
+    * Should be able to use in the `limit` api usage
+    """
+    with limit(subfunction) as rf:
+        this_process_id = os.getpid()
+        other_process_id = rf()
+
+    assert this_process_id != other_process_id
+
+
+def test_bad_func_arg() -> None:
+    """
+    Expects
+    -------
+    * Should raise an Error about a none callable bad memory limit
+    """
+    with pytest.raises(ValueError, match=r"func"):
+        Pynisher(func=None)  # type: ignore
+
+
+@pytest.mark.parametrize("memory", [-1, 0])
+def test_bad_memory_arg(memory: int) -> None:
+    """
+    Expects
+    -------
+    * Should raise an Error about a bad memory limit
+    """
+    with pytest.raises(ValueError, match=r"memory"):
+        Pynisher(subfunction, memory=memory)
+
+
+@pytest.mark.parametrize("cpu_time", [-1, 0])
+def test_bad_cpu_time_arg(cpu_time: int) -> None:
+    """
+    Expects
+    -------
+    * Should raise an Error about a bad cpu_time limit
+    """
+
+    def _f() -> None:
+        pass
+
+    with pytest.raises(ValueError, match=r"cpu_time"):
+        Pynisher(_f, cpu_time=cpu_time)
+
+
+@pytest.mark.parametrize("wall_time", [-1, 0])
+def test_bad_wall_time_arg(wall_time: int) -> None:
+    """
+    Expects
+    -------
+    * Should raise an Error about a bad wall_time limit
+    """
+
+    def _f() -> None:
+        pass
+
+    with pytest.raises(ValueError, match=r"wall_time"):
+        Pynisher(_f, wall_time=wall_time)
+
+
+def test_bad_context_arg() -> None:
+    """
+    Expects
+    -------
+    * Should raise an Error about a bad context arg
+    """
+
+    def _f() -> None:
+        pass
+
+    with pytest.raises(ValueError, match=r"context"):
+        Pynisher(_f, context="bad arg")
+
+
+def test_no_raise_gets_empty() -> None:
```

Comment:
Why doesn't this test get a context fixture?

Reply:
I could throw it in, it's not really dependent on contexts. I'll just throw in contexts for all tests then because why not.

Reply:
> I'll just throw in contexts for all tests then because why not.

That would be great!